cosima_cookbook/database.py

```diff
@@ -578,86 +578,93 @@ def build_index(
+        if len(files) > 0:
+            if len(expt.ncfiles) > 0:
+                # Only pass files that are not already in DB
+                files = {
+                    f.ncfile
+                    for f in session.query(NCFile)
+                    .with_parent(expt)
+                    .filter(NCFile.ncfile.notin_(files))
+                }
 
-        # if everything went smoothly, commit these changes to the database
-        session.commit()
+            indexed += index_experiment(files, session, expt, client)
+
+            # if everything went smoothly, commit these changes to the database
+            session.commit()
 
     return indexed
 
 
-def prune_files(expt, session, files, delete=True):
-    """Delete or mark as not present the database entries for specified
-    files objects within the given experiment
+def _prune_files(expt, session, files, delete=True):
+    """Delete or mark as not present the database entries that are
+    not present in the list of files
     """
-    for f in files:
+    if len(expt.ncfiles) == 0:
+        # No entries in DB, special case just return as there is nothing
+        # to prune and trying to do so will raise errors
+        return
+
+    # Missing are physically missing from disk, or where marked as not
+    # present previously. Can also be a broken file which didn't index
+    missing_ncfiles = (
+        session.query(NCFile)
+        .with_parent(expt)
+        .filter(NCFile.ncfile.notin_(files) | (NCFile.present == False))
+    )
+    if missing_ncfiles.first() is not None:
         if delete:
-            session.delete(f)
+            missing_ncfiles.delete(synchronize_session=False)
         else:
-            f.present = False
+            missing_ncfiles.update({NCFile.present: False}, synchronize_session=False)
+
+        session.commit()
 
 
-def prune_experiment(experiment, session, delete=True):
+def prune_experiment(experiment, session, delete=True, followsymlinks=False):
     """Delete or mark as not present the database entries for files
     within the given experiment that no longer exist or were broken at
-    index time.
+    index time. Experiment can be either an NCExperiment object, or the
+    name of an experiment available within the current session.
     """
 
-    expt = (
-        session.query(NCExperiment)
-        .filter(NCExperiment.experiment == experiment)
-        .one_or_none()
-    )
-
-    if not expt:
-        print("No such experiment: {}".format(experiment))
-        return
-
-    files = {f for f in expt.ncfiles if not f.present or not f.ncfile_path.exists()}
+    if isinstance(experiment, NCExperiment):
+        expt = experiment
+        experiment = expt.experiment
+    else:
+        expt = (
+            session.query(NCExperiment)
+            .filter(NCExperiment.experiment == experiment)
+            .one_or_none()
+        )
+        if not expt:
+            print("No such experiment: {}".format(experiment))
+            return
```

Comment:
I'm probably on the `git blame` for adding this originally, but this should probably raise a `RuntimeError` instead, right?

Reply:
Made the message a `logging` one, but just realised I should have added it to the exception. Will change.